cmake 3.15+ MSVC runtime library fix

mattiasljungstrom opened the pull request:
ok, second try!

mattiasljungstrom commented:
Another option would be to remove this fips option completely and let people setup the flags on their own. The default would be /MD. (Your default in fips was /MT.)

floooh commented:
Hmm I'm still having trouble with C++ libraries. I'm getting linker errors like this:

```
unresolved external symbol __imp__invalid_parameter referenced in function "public: unsigned short & __cdecl std::vector<unsigned short,class ozz::StdAllocator<unsigned short> >::operator[](unsigned __int64)" (??A?$vector@GV?$StdAllocator@G@ozz@@@std@@QEAAAEAG_K@Z)
6>ozzanim.lib(ozz_animation.obj) : error LNK2001: unresolved external symbol __imp__invalid_parameter
```

Looking into the Visual Studio settings dialog I can see that both the Debug and Release build config link with the Release runtime libraries (both configs use "Multi-threaded DLL (/MD)".  without the PR applied, the setting is "Multi-threaded (/MT)" for release mode, and "Multi-threaded Debug (/MTd)" for Debug mode. It's as if the cmake settings are not applied:

![image](https://user-images.githubusercontent.com/1699414/153633865-961c212c-9a04-41d8-bec1-efb3fb51b684.png)

![image](https://user-images.githubusercontent.com/1699414/153634301-dca5661c-b58b-4fce-bdee-d6021dead108.png)

This is with the sokol-samples project, sapp-d3d11-win64-vstudio-debug build config, cmake 3.22.2, and Visual Studio Community 2019.

floooh commented:
...oh, there's an important caveat in the CMake docmentation:

Note This variable has effect only when policy [CMP0091](https://cmake.org/cmake/help/git-stage/policy/CMP0091.html#policy:CMP0091) is set to NEW prior to the first [project()](https://cmake.org/cmake/help/git-stage/command/project.html#command:project) or [enable_language()](https://cmake.org/cmake/help/git-stage/command/enable_language.html#command:enable_language) command that enables a language using a compiler targeting the MSVC ABI.

Guess we should wait until cmake switches the default for the policy from OLD to NEW(?)

mattiasljungstrom commented:
Hmmm.... With CMake 3.22.2 and VS2019 latest update it works with my patch, but I haven't tested lots of different setups of course. 

If you set cmake_minimum_required(VERSION 3.22) it seems like the new behavior is used? At least I get /MD setup in my VS projects and I need to manually add CMAKE_MSVC_RUNTIME_LIBRARY in my main cmake file.

mattiasljungstrom commented:
If you want to wait with this I will survive with my own changes locally.

floooh commented:
> If you set cmake_minimum_required(VERSION 3.22...

Ah right, that would be the better solution. I wonder if this version check:

```
if (${CMAKE_VERSION} VERSION_GREATER_EQUAL 3.15)
```
...can be replaced with a check that tests the policy or the version that's set with "cmake_minimum_required()", then it should work automatically with old and new cmake files and also old cmake files on new cmake versions.

mattiasljungstrom commented:
Good point, I've pushed a change that checks if the policy is active instead of looking at the CMake version. I've not tested this on older CMake versions, but it works for both cases on CMake 3.22 for me.

mattiasljungstrom commented:
Btw, the CMake documentation seems to be incorrect, it says that this policy defaults to OLD but in my project it defaults to NEW with CMake 3.22.

floooh commented:
Very weird, I'm now getting this when running "fips gen" with build config ```sapp-d3d11-win64-vstudio-debug``` in the sokol samples. I'll investigate.

```
CMake Error at E:/projects/fips/cmake-toolchains/windows.cmake:63 (if):
  if given arguments:

    "STREQUAL" "NEW"

  Unknown arguments specified
Call Stack (most recent call first):
  E:/projects/fips/cmake/fips.cmake:166 (include)
  CMakeLists.txt:22 (fips_setup)


-- Configuring incomplete, errors occurred!
```

floooh commented:
> Btw, the CMake documentation seems to be incorrect, it says that this policy defaults to OLD but in my project it defaults to NEW with CMake 3.22.

I guess you have a cmake_policy() hidden somewhere, because otherwise cmake_policy(GET ...) would return an empty variable.

mattiasljungstrom commented:
The variable needed to be inside a string "", otherwise there was an error when it was empty. Fixed that and squashed the commits. Tested this on your setup now as well with sokol-samples.

mattiasljungstrom commented:
If I run sokol-samples as is (cmake_minimum_required(VERSION 3.2)) I get empty string "", but if I change cmake_minimum_required to 3.15 I get "NEW".

floooh commented:
Ok this works great :)

I tested on sokol-samples without changing the cmake files, both for Win32 and UWP builds.

This policy behaviour in cmake is still confusing... just one of many tiny things unfortunately.
